feat(api): Flex Stacker Module Support for EVT

ahiuchingau opened the pull request:
# Overview
Covers EXEC-967, EXEC-965, EXEC-946, EXEC-1078

This PR introduces the .store(), .retrieve(), and .load_labware_to_hopper(...) commands. These commands allow the declaration of labware inside the hopper, retrieval of a handleable labware core from the stacker, and storage of labware into the stacker.

## Test Plan and Hands on Testing
- [x] The following protocol should pass app analysis and execute as expected on a Flex:
```
# Make sure apiLevel is high enough that this protocol will be run through Protocol Engine.
requirements = {"robotType": "Flex", "apiLevel": "2.23"}


def run(protocol):
    stacker_c = protocol.load_module("flexStackerModuleV1", "C4")
    stacker_d = protocol.load_module("flexStackerModuleV1", "D4")
    stacker_c.load_labware_to_hopper(
        "opentrons_flex_96_tiprack_1000ul", 2)
    stacker_d.load_labware_to_hopper(
        "corning_96_wellplate_360ul_flat", 2)

    """You can now load labware on the stacker slot as a staging slot."""
    stacker_c.enter_static_mode()
    tiprack_d = stacker_c.load_labware("opentrons_flex_96_tiprack_1000ul")
    protocol.move_labware(tiprack_d, "D3", use_gripper=True)
    stacker_c.exit_static_mode()
    
    tiprack = stacker_c.retrieve()
    protocol.move_labware(tiprack, "C2", use_gripper=True)

    plate = stacker_d.retrieve()
    protocol.move_labware(plate, "D2", use_gripper=True)

    protocol.move_labware(tiprack, stacker_c, use_gripper=True)
    stacker_c.store(tiprack)

    protocol.move_labware(plate, stacker_d, use_gripper=True)
    stacker_d.store(plate)
```

## Changelog


## Review requests
Outside of the known unintentional allowance of loading multiple stackers in the same slot (due to skipping the modules map on the deck conflict checker), this EVT version is meant to track labware through an internally managed labware list in the modules substate. Future versions may not reflect that practice.

Are there problems with the remaining approaches in this PR, whose purpose is to enable to use of the Flex stacker alongside all 12 base deck slots for nominal protocol operation?

## Risk assessment
Low - Mostly introduction of new behavior with limited overlap to existing features. One existing risk is that in this EVT build multiple Flex Stackers can be loaded to the same slot due to the Stacker skipping the deck map conflict checker for the modules map.

ahiuchingau committed:
update load_labware to handle loading labware to stacker hopper

ahiuchingau committed:
add flex_stacker.load_labware_to_hopper

ahiuchingau committed:
fixup! add flex_stacker.load_labware_to_hopper

ahiuchingau committed:
add unit tests

ahiuchingau committed:
fix circular import

ahiuchingau committed:
use correct fixture location

ahiuchingau committed:
enable retrieve command

ahiuchingau committed:
add store command capability as well

ahiuchingau committed:
fix some tests

ahiuchingau committed:
got it to work

ahiuchingau committed:
Merge branch 'edge' into EXEC-1078

ahiuchingau committed:
fix failed tests

jbleon95 commented on api/src/opentrons/protocol_api/core/legacy/legacy_protocol_core.py:
```diff
@@ -524,6 +524,19 @@ def load_lid_stack(
+        module_core: legacy_module_core.LegacyModuleCore,
+        load_name: str,
+        quantity: int,
+        label: str | None = None,
```
Let's have this (and the ones below) `Optional[str]` for consistency

CaseyBatten commented on api/src/opentrons/protocol_engine/commands/load_labware.py:
```diff
@@ -145,9 +156,24 @@ async def execute(
-            params.location
-        )
+        verified_location: LabwareLocation
+        if (
+            self._is_loading_to_module(
+                params.location, ModuleModel.FLEX_STACKER_MODULE_V1
+            )
+            and not self._state_view.modules.get_flex_stacker_substate(
+                params.location.moduleId
+            ).in_static_mode
+        ):
```
Looks good for EVT purposes, eventually we might want to de-coordinate these things if we're able to source the labware locations to a true stack in the tower.

CaseyBatten commented on api/src/opentrons/protocol_engine/commands/load_labware.py:
```diff
@@ -186,12 +212,20 @@ async def execute(
+                loaded_labware.definition
+            )
+
+        if self._is_loading_to_module(
+            params.location, ModuleModel.FLEX_STACKER_MODULE_V1
+        ):
+            state_update.load_flex_stacker_hopper_labware(
+                module_id=params.location.moduleId,
+                labware_id=loaded_labware.labware_id,
+            )
```
Makes sense for this build.

jbleon95 commented on api/src/opentrons/protocol_api/protocol_context.py:
```diff
@@ -862,7 +866,11 @@ def load_module(
+                  .. versionchanged:: 2.23
+                    Added ``FlexStackerModuleContext`` return value.
         """
+        flex_stacker_valid_since = APIVersion(2, 23)
```
If this is a constant can we make it ALL_CAPS (and maybe move it to where we keep other version gates)

CaseyBatten commented on api/src/opentrons/protocol_engine/state/module_substates/flex_stacker_substate.py:
```diff
@@ -15,3 +22,34 @@ class FlexStackerSubState:
+            # this is a temporary solution to enable evt testing
+            if isinstance(lw_change, FlexStackerLoadHopperLabware):
+                # for manually loading labware in the stacker
+                new_labware_ids.append(lw_change.labware_id)
+            elif isinstance(lw_change, FlexStackerRetrieveLabware):
+                new_labware_ids.remove(lw_change.labware_id)
+            elif isinstance(lw_change, FlexStackerStoreLabware):
```
I'm cool with the way these lists work for this build. Eventually we probably want to transition to letting the labware state store manage this by having them be part of a location hierarchy/stack. This can come later though as theres more architecture that would need changing to allow that to work.

CaseyBatten commented on api/src/opentrons/protocol_engine/state/modules.py:
```diff
@@ -1319,9 +1331,9 @@ def ensure_and_convert_module_fixture_location(
-            # only allowed in column 4
-            assert deck_slot.value[-1] == "4"
-            return f"flexStackerModuleV1{deck_slot.value}"
+            # loaded to column 3 but the addressable area is in column 4
+            assert deck_slot.value[-1] == "3"
+            return f"flexStackerModuleV1{deck_slot.value[0]}4"
```
This is fine for the time being, but eventually we'll want to rethink the architecture below this to allow the module to actually load into D4 instead of a fake D4.

CaseyBatten reviewed (approved):
Looks solid, added some notes below on places for future improvements to allow us to consolidate this down to the labware state store instead of spread across the module substate. A true labware stack would take quite a bit of rearchitecting elsewhere, but might be worthwhile in the long run.

SyntaxColoring commented on api/src/opentrons/protocol_api/core/engine/module_core.py:
```diff
@@ -700,8 +700,17 @@ class FlexStackerCore(ModuleCore, AbstractFlexStackerCore):
+
     def retrieve(self) -> None:
-        """Retrieve a labware from the bottom of the Flex Stacker's stack."""
+        """Retrieve a labware id from the bottom of the Flex Stacker's stack."""
```
Confused about this docstring change. It doesn't look like we are returning an ID?

SyntaxColoring commented on api/src/opentrons/protocol_api/core/engine/protocol.py:
```diff
@@ -373,6 +374,37 @@ def load_lid(
+        label: Optional[str] = None,
+        namespace: Optional[str] = None,
+        version: Optional[int] = None,
+        lid: Optional[str] = None,
```
Nitpick: I would not default these arguments to `= None`. If the caller doesn't provide these arguments, it is almost certainly a bug, so we should stop them from being able to do that. And there is only one caller outside of tests, so that shouldn't be too verbose.

SyntaxColoring commented on api/src/opentrons/protocol_api/module_contexts.py:
```diff
@@ -1112,21 +1112,90 @@ class FlexStackerContext(ModuleContext):
+        version: Optional[int] = None,
+        lid: Optional[str] = None,
+    ) -> None:
+        """Load one or more labware onto the flex stacker.
+
+        The parameters of this function behave like those of
+        :py:obj:`ProtocolContext.load_labware` (which loads labware directly
+        onto the deck). Note that the parameter ``name`` here corresponds to
+        ``load_name`` on the ``ProtocolContext`` function.
+
+        :returns: The initialized and loaded labware object.
+
+        .. versionadded:: 2.23
+            The *label,* *namespace,* and *version* parameters.
+        """
```
1. There is no `name` parameter like the docstring describes. Is that sentence left over from a copy-paste?
2. I don't think this `.. versionadded` block is meaningful, because the entire *function* was added in 2.23. I think we should just delete it?

SyntaxColoring commented on api/src/opentrons/protocol_api/module_contexts.py:
```diff
@@ -1112,21 +1112,90 @@ class FlexStackerContext(ModuleContext):
+            lid=lid,
+        )
+
+    def enter_static_mode(self) -> None:
+        """Enter static mode.
+
+        In static mode, the Flex Stacker will not move labware between the hopper and
+        the deck. The stacker can be used as a staging slot area.
+
+        .. versionadded:: 2.23
+        """
+        self._core.set_static_mode(static=True)
+
+    def exit_static_mode(self) -> None:
+        """End static mode.
+
+        In static mode, the Flex Stacker will not move labware between the hopper and
+        the deck. This is useful for debugging and manual operation.
+
+        .. versionadded:: 2.23
+        """
+        self._core.set_static_mode(static=False)
```
These should have a `@requires_version` decorator, and I think we can remove the `.. versionadded` blocks because `@requires_version` will add them automatically.

SyntaxColoring replied:
Looks like these docstrings are slightly inconsistent in their explanations of what static mode is.

SyntaxColoring commented on api/src/opentrons/protocol_api/module_contexts.py:
```diff
@@ -1112,21 +1112,90 @@ class FlexStackerContext(ModuleContext):
+    def retrieve(self) -> Labware:
         """Release and return a labware at the bottom of the labware stack."""
         self._core.retrieve()
+        labware_core = self._protocol_core.get_labware_on_module(self._core)
+        assert labware_core is not None, "Retrieve failed to return labware"
```
If I'm reading this correctly, I don't think an `assert` is appropriate here?

If you:

1. Load a stacker
2. *Do not load any labware into it*
3. Call `retrieve()`

Then it will be the case that `labware_core` is `None`, right? So this condition isn't a bug in Opentrons code—it's a problem in the user's protocol that we need to signal with a proper `EnumeratedError`.

ahiuchingau replied:
The engine retrieve command should have raised an error before we get to this point. I wasn't sure if I should let the engine handle this only or handle it in the module context as well. From your comment, I think the answer is both?

SyntaxColoring replied:
Aah, so any time this `assert` would have raised, you're expecting the `self._core.retrieve()` call a couple lines above to have raised, anyway?

Okay, that makes sense, and no, I don't think you need to change this `assert` in that case. I would probably leave a comment explaining this, though, or maybe elaborate on it in the `assert` message.

jbleon95 commented on api/src/opentrons/protocol_api/core/engine/protocol.py:
```diff
@@ -373,6 +374,37 @@ def load_lid(
+        """Load one or more labware with or without a lid to the flex stacker hopper."""
+        assert (
+            isinstance(module_core, ModuleCore)
+            and module_core.MODULE_TYPE == ModuleType.FLEX_STACKER
```
Wouldn't be virtue of being a `FLEX_STACKER`, it would be guaranteed to not be a `NonConnectedModuleCore`?

SyntaxColoring commented on api/src/opentrons/protocol_api/module_contexts.py:
```diff
@@ -1112,21 +1112,90 @@ class FlexStackerContext(ModuleContext):
 
         :param labware: The labware object to store.
         """
+        assert labware._core is not None
```
Is this doing anything? It looks like `labware._core` is statically typechecked to never be `None`.

ahiuchingau replied:
This is so that user can call this function:
```
stacker.store(plate)
```
In a refactor, we will add the labware id as a store command param so we can make sure the labware has already been loaded on the stacker slot.

SyntaxColoring replied:
Right but I mean, is the `assert` statement on line 1200 doing anything?

SyntaxColoring commented on api/src/opentrons/protocol_api/protocol_context.py:
```diff
@@ -901,7 +920,11 @@ def load_module(
         )
         if isinstance(deck_slot, StagingSlotName):
-            raise ValueError("Cannot load a module onto a staging slot.")
+            # flex stacker modules can only be loaded into staging slot inside a protocol
```
Is this the right layer to enforce this? We don't want it in the Protocol Engine `loadModule` command?

ahiuchingau replied:
So we had to do this because the flex stacker is the very first module that provides an addressable area (deck column 4) in a different slot than its cutout mount (column 3). 

Technically we are still loading the module to column 3 of the deck in the engine (because the engine expects only deck slots). But to users, they wouldn't care which deck cutout the module is being loaded into. It would be weird have user load the module into column 3 in the protocol when the observable addressable area is in column 4. 

We did this here so that we can keep the same engine behavior for the flex stacker for fast development. In a future refactor, we should use addressable areas as the load location on the context level, instead of physical deck slots.

SyntaxColoring replied:
OK, I'm only tenuously following, but I can catch up on this later. It doesn't seem like anything that should block this PR. Thanks for explaining!

SyntaxColoring commented on api/src/opentrons/protocol_engine/commands/flex_stacker/configure.py:
```diff
@@ -0,0 +1,76 @@
+
+
+class Configure(BaseCommand[ConfigureParams, ConfigureResult, ErrorOccurrence]):
+    """A command to Configure a labware from a Flex Stacker."""
```
Is this docstring correct? We're configuring the stacker itself, not its labware, right?

SyntaxColoring commented on api/src/opentrons/protocol_engine/commands/flex_stacker/retrieve.py:
```diff
@@ -43,19 +52,54 @@ def __init__(
-            await stacker.dispense_labware(labware_height=50.0)
+        try:
+            self._state_view.labware.raise_if_labware_in_location(stacker_loc)
+        except Exception as e:
```
Can we specifically catch `LocationIsOccupiedError` instead of `Exception`?

SyntaxColoring commented on api/src/opentrons/protocol_engine/commands/flex_stacker/retrieve.py:
```diff
@@ -43,19 +52,54 @@ def __init__(
+        try:
+            self._state_view.labware.raise_if_labware_in_location(stacker_loc)
+        except Exception as e:
+            raise CannotPerformModuleAction(
+                f"Cannot retrieve a labware from Flex Stacker if the carriage is occupied: {e}"
+            )
```
1. As a general practice, I think we should avoid doing `raise NewException(f"blah {old_exception}")`, where we blindly stringify the old exception and incorporate it into the new one's message.

   `{e}` is not generally guaranteed to make sense when you concatenate it like this. So this pattern only works in the cases where the exception type is constrained to something that you *know* will stringify nicely as part of its API contract, like `JSONDecodeError`. In all other cases, it risks concatenating into a confusing string because of too little information (like `KeyError`) or redundant information (like a lot of our command error handling in Protocol Engine).

   Instead, we should either:

   * Just have a static message here and delete `{e}`. (I would do this.)
   * Replace `catch Exception` with a specific exception type, and replace `{e}` with something more specific and explicit like `{e.conflicting_labware}`.

5. For a correct machine- and developer-readable error tree, we should do `raise ... from e` and `CannotPerformModuleAction(..., wrapping=[e])`.

sfoster1 commented on api/src/opentrons/protocol_api/core/engine/module_core.py:
```diff
@@ -700,8 +700,17 @@ class FlexStackerCore(ModuleCore, AbstractFlexStackerCore):
     _sync_module_hardware: SynchronousAdapter[hw_modules.FlexStacker]
 
+    def set_static_mode(self, static: bool) -> None:
+        """Set the Flex Stacker's static mode."""
```
what's a static mode?

SyntaxColoring reviewed (commented):
I'm about halfway through and this all looks good so far, thanks. Just various small things.

sfoster1 reviewed (approved):
Looks like a great first pass!

SyntaxColoring commented on api/src/opentrons/protocol_engine/commands/flex_stacker/store.py:
```diff
@@ -44,17 +46,40 @@ def __init__(
-        stacker = self._equipment.get_module_hardware_api(stacker_substate.module_id)
+        stacker_hw = self._equipment.get_module_hardware_api(stacker_state.module_id)
+
+        try:
+            lw_id = self._state_view.labware.get_id_by_module(params.moduleId)
+        except Exception:
+            raise CannotPerformModuleAction(
+                "Cannot store labware if Flex Stacker carriage is empty"
+            )
+
```
Can we specifically catch `LabwareNotLoadedOnModuleError` instead of `Exception`?

ahiuchingau committed:
address pr comments

SyntaxColoring commented on api/src/opentrons/protocol_engine/commands/load_labware.py:
```diff
@@ -145,9 +156,24 @@ async def execute(
+                params.location.moduleId
+            ).in_static_mode
+        ):
+            # labware loaded to the flex stacker hopper is considered offdeck. This is
+            # a temporary solution until the hopper can be represented as non-addressable
+            # addressable area in the deck configuration.
```
> non-addressable addressable area in the deck configuration

Is that the direction we're headed? I thought we wanted something that's like `OFF_DECK` but not `OFF_DECK`?

github-actions[bot] committed:
fix(analyses-snapshot-testing): heal EXEC-1078 snapshots (#17304)

This PR was requested on the PR
https://github.com/Opentrons/opentrons/pull/17300

ahiuchingau committed:
catch proper exceptions

SyntaxColoring reviewed (approved):
Thank you!

ahiuchingau committed:
add a comment
